### ISSUE TITLE
Linkerd continues to talk to old endpoint after Kubernetes deployment

### ISSUE THREAD

**Issue by dpetersen** [linkerd/linkerd#1730]
Issue Type:

- [x] Bug report
- [ ] Feature request

**What happened**:

I upgraded our Linkerd from 0.8.6 (I know, ancient, but we had no complaints) to 1.3.3. We are running a Kubernetes cluster, and this Linkerd receives traffic from the Internet via an ELB and routes to the appropriate services in various staging environments. I updated our configuration, and everything appeared to be working fine.

Here is our configuration. This upgraded linkerd is named `beyond-thunderdome` because the old one was `thunderdome` and it all makes perfect sense if you've watched Mad Max:
```
    admin:
      port: 9990
      # because it's getting hit by k8s for health checks
      ip: 0.0.0.0
    usage:
      enabled: false
    routers:
      - label: in-beyond-thunderdome
        protocol: http
        #httpAccessLog: /dev/stdout # not in production, por favor
        servers:
          - port: 8080
            # on purpose, in-pod nginx sidecar will be able to proxy to
            # localhost
            ip: 127.0.0.1 
            # this clears out linkerd headers that let you manipulate cluster
            # routing. It is ABSOLUTELY CRITICAL that we don't accept these
            # headers from the internet.
            clearContext: true
        interpreter:
          kind: io.l5d.k8s.configMap
          experimental: true # sure, why not
          name: linkerd-dtabs
          filename: beyond-thunderdome
          namespace: default
    namers:
      - kind: io.l5d.k8s
        host: 127.0.0.1
        port: 8001
    telemetry:
      - kind: io.l5d.prometheus
        path: /admin/metrics/prometheus
        prefix: linkerd_
      - kind: io.l5d.recentRequests
        sampleRate: 0.0001
        capacity: 20
```

slimmed-down dtab ConfigMap looks like this:

```apiVersion: v1
kind: ConfigMap
metadata:
  name: linkerd-dtabs
data:
  beyond-thunderdome: |-
    /host/com/example/integrations-dashboard => /#/io.l5d.k8s/integrations-staging/http/integrations-dashboard;
    /svc => /$/io.buoyant.http.domainToPathPfx/host ;
```

A couple of successful deploys were done after the upgrade, but I was notified that one of the services was down, returning 502s after a deployment. This is all low-traffic staging environments, so both the linkerd and the target service are scaled to 1. The service is called `integrations-dashboard` in the namespace `integrations-staging`. For the sake of this issue, the address is `integrations-dashboard.example.com`.

Our logs were full of errors like:

```
E 1206 01:05:20.259 UTC THREAD28: service failure: Failure(No route to host: /100.96.65.31:3000 at remote address: /100.96.65.31:3000. Remote Info: Not Available, flags=0x09) with RemoteInfo -> Upstream Address: Not Available, Upstream id: Not Available, Downstream Address: /100.96.65.31:3000, Downstream label: #/io.l5d.k8s/integrations-staging/http/integrations-dashboard, Trace Id: d1d6dc90db15ca53.d1d6dc90db15ca53<:d1d6dc90db15ca53
```

Inspecting the Kubernetes service, the address `100.96.65.31` was incorrect. I can't say for sure, but I believe this was the internal address of the service prior to their deployment. Interestingly, if I went to the Linkerd admin console dtab playground and put `/svc/integrations-staging.example.com` it returned the correct internal address. The Prometheus metrics show a failure accrual removal that corresponded to the exact time they performed a deploy.

We tried another deployment and it didn't fix anything. I enabled verbose logging, deployed again, and saw this:

```
D 1206 03:04:31.913 UTC THREAD28: k8s ns integrations-staging service integrations-dashboard modified endpoints
D 1206 03:04:49.014 UTC THREAD28: k8s ns integrations-staging service integrations-dashboard modified endpoints                                                                   
D 1206 03:04:49.015 UTC THREAD28: k8s ns integrations-staging service integrations-dashboard added Endpoint(/100.96.63.61,Some(ip-172-20-100-132.us-west-2.compute.internal))     
D 1206 03:04:49.113 UTC THREAD28: k8s ns integrations-staging service integrations-dashboard modified endpoints
D 1206 03:04:49.114 UTC THREAD28: k8s ns integrations-staging service integrations-dashboard removed Endpoint(/100.97.4.48,Some(ip-172-20-61-49.us-west-2.compute.internal))      
E 1206 03:05:02.996 UTC THREAD31: service failure: Failure(connection timed out: /100.96.65.31:3000 at remote address: /100.96.65.31:3000. Remote Info: Not Available, flags=0x09) with RemoteInfo -> Upstream Address: Not Available, Upstream id: Not Available, Downstream Address: /100.96.65.31:3000, Downstream label: #/io.l5d.k8s/integrations-staging/http/integrations-dashboard, Trace Id: d1d6dc90db15ca53.d1d6dc90db15ca53<:d1d6dc90db15ca53                                                                                             
I 1206 03:05:10.192 UTC THREAD28: FailureAccrualFactory marking connection to "#/io.l5d.k8s/integrations-staging/http/integrations-dashboard" as dead. Remote Address: Inet(/100.96.65.31:3000,Map(nodeName -> ip-172-20-63-116.us-west-2.compute.internal))
E 1206 03:05:10.194 UTC THREAD28: service failure: Failure(connection timed out: /100.96.65.31:3000 at remote address: /100.96.65.31:3000. Remote Info: Not Available, flags=0x09) with RemoteInfo -> Upstream Address: Not Available, Upstream id: Not Available, Downstream Address: /100.96.65.31:3000, Downstream label: #/io.l5d.k8s/integrations-staging/http/integrations-dashboard, Trace Id: d1d6dc90db15ca53.d1d6dc90db15ca53<:d1d6dc90db15ca53
```

It seems like it is seeing the changing endpoints correctly, but it's fixated on talking to the old endpoint and refuses to give up and move on.

I tried to troubleshoot the problem, but eventually to get the service back up I just restarted linkerd and everything immediately started working.

**What you expected to happen**:

I expected it to remove the old endpoint from its list when the Kubernetes service told it to.

**How to reproduce it (as minimally and precisely as possible)**:

I wish I knew. I will endeavor to try and reproduce the problem. Like I said, we've had a couple of identical deploys occur without an issue.

**Anything else we need to know?**:

Nope. Thanks for reading this enormous ticket.

**Environment**:
- Linkerd 1.3.3
- Kubernetes 1.7.2
- Cloud provider or hardware configuration: AWS

**Comment by hawkw** [comment]
Hi @dpetersen, thanks for _writing_ this enormous ticket! I really appreciate seeing the detailed description of the problem. 

We'll look into this and get back to you if I need any more information.

**Comment by hawkw** [comment]
Okay, I've reproduced this issue with a similar configuration to yours on Linkerd 1.3.3. We're currently trying to determine the cause of this issue.

**Comment by hawkw** [comment]
@dpetersen, what sequence of commands is actually being used to do the deployment?

**Comment by dpetersen** [comment]
We have some CI/CD automation that is updating the Deployment object using the K8s go client library. We've had a dozen deploys go through successfully today without any failures, and we're still running 1.3.3. I was hoping to have it happen again since I still have verbose logging turned on.

I can provide more details of the precise code if you'd like. We also have Prometheus metrics for this instance, including for last night when we had our issue. I can put some of those in this ticket if it helps. I mostly looked at `rt:client:failure_accrual:removals` and `rt:service:failures:rejected:com_twitter_finagle_Failure:com_twitter_finagle_ConnectionFailedException`, since those two jumped out at me while I was reading the metrics endpoint during troubleshooting.

**Comment by hawkw** [comment]
As much metrics data as you can provide would be highly appreciated!

**Comment by hawkw** [comment]
Morning @dpetersen, something else that could be useful for debugging this issue is if you could set the log level for the `k8s` scope to `TRACE` rather than `DEBUG`. You can do this by opening the Linkerd admin interface, clicking on "logging" in the top menu bar, and scrolling down to find `k8s`. You _probably_ don't want to set the `TRACE` log level globally - it'll make your logs very verbose.

**Comment by dpetersen** [comment]
That was one of the first things I did after we had our issue. I was watching the logs while our service was still down. Some of the logs I posted above are with k8s at a trace level while we were doing subsequent deploys trying to fix the service.

One thing I did not post is that the logs are full of this:
```
D 1207 19:33:32.593 UTC THREAD28: k8s returned 'too old resource version' error with incorrect HTTP status code, restarting watch
D 1207 19:33:32.593 UTC THREAD28: k8s restarting watch on /api/v1/watch/namespaces/content-staging/endpoints/assets, resource version None was too old
```
I'm assuming that's normal behavior.

Another change I didn't think of is that we upgraded the `buoyantio/kubectl` sidecar pod from `v1.4.0` to `v1.6.2` (the most recent published version), although our k8s version is actually `1.7.2`. Those logs will occasionally spew:
```
I1207 18:59:35.249296       1 logs.go:41] httputil: ReverseProxy read error during body copy: context canceled
I1207 18:59:35.254915       1 logs.go:41] httputil: ReverseProxy read error during body copy: context canceled
I1207 19:29:42.662030       1 logs.go:41] httputil: ReverseProxy read error during body copy: context canceled
I1207 19:29:42.680564       1 logs.go:41] httputil: ReverseProxy read error during body copy: context canceled
I1207 19:29:42.697225       1 logs.go:41] httputil: ReverseProxy read error during body copy: context canceled
I1207 19:29:42.707671       1 logs.go:41] httputil: ReverseProxy read error during body copy: context canceled
I1207 19:29:42.725691       1 logs.go:41] httputil: ReverseProxy read error during body copy: context canceled
I1207 19:29:42.735628       1 logs.go:41] httputil: ReverseProxy read error during body copy: context canceled
I1207 19:47:35.331696       1 logs.go:41] httputil: ReverseProxy read error during body copy: context canceled
I1207 19:47:35.350363       1 logs.go:41] httputil: ReverseProxy read error during body copy: context canceled
I1207 19:47:35.359299       1 logs.go:41] httputil: ReverseProxy read error during body copy: context canceled
```
From the timestamps, it's not constant but just a few times an hour. I looked at that during our issue the other night and none of the timestamps seemed to correspond to the same time as our outage. Our production linkerd is still running old linkerd with the `v1.4.0` against that Kubernetes `1.7.2` and has no such errors, but that might just be down to logging changes in kubectl.

I'm still digging into the metrics. Since it's time series I can't just dump numbers in here and provide anything useful, so I'm trying to look at all the metrics for the time period of the event to see if I find anything that looks weird. So far the only curious thing I found was that during the window of the outage, the `rt:service:request_latency_ms` metric for that particular service capped out at 1s and never went above (or much below). Here is a graph from +/-1hr of the outage.

<img width="690" alt="screenshot" src="https://user-images.githubusercontent.com/408/33735855-6ba6b7fe-db45-11e7-8104-735eadc988e9.png">

I don't know what it means, it just stuck out to me as looking kind of funny. I will continue looking at the metrics.

I should also point out that I am *still* running the same config and we've had over 30 deploys in the last two days that have not had the issue. As far as I know, there was nothing special about that deploy (that app has been deployed a few times since).

**Comment by dpetersen** [comment]
For completeness sake, here are the graphs from the same time period for `linkerd_rt:client:failure_accrual:removed_for_ms` and `linkerd_rt:service:failures:rejected:com_twitter_finagle_Failure:com_twitter_finagle_ConnectionFailedException` that I referred to earlier.

<img width="579" alt="screenshot" src="https://user-images.githubusercontent.com/408/33736430-6ad60aee-db47-11e7-8a56-b6018d703deb.png">

**Comment by wmorgan** [comment]
Thank you for all the detail. Very helpful. We continue to investigate.

**Comment by siggy** [comment]
@dpetersen We've just released 1.3.4 with some fixes in the k8s namer. Would you mind testing with this release?

**Comment by dpetersen** [comment]
I will deploy it now. I was just about to tell you that we haven't had a recurrence of this issue in the last two weeks, but I just checked my dashboard and we actually had a failed deployment last night. I went in the slack channel of the team involved and they were trying to figure out why their service was unresponsive, but it just sort of healed itself after a few hours. I went into my logs, and here are the relevant lines. I have k8s at a trace log level, unlike my previous log posting. Unfortunately, there isn't much new information here:

```
{"log":"D 1218 04:35:22.595 UTC THREAD31: k8s returned 'too old resource version' error with incorrect HTTP status code, restarting watch\n","stream":"stderr","time":"2017-12-18T04:35:22.597129146Z"}
{"log":"I 1218 04:35:26.607 UTC THREAD23: FailureAccrualFactory marking connection to \"#/io.l5d.k8s/integrations-staging/http/inventory\" as dead. Remote Address: Inet(/100.97.4.75:3000,Map(nodeName -\u003e ip-172-20-61-49.us-west-2.compute.internal))\n","stream":"stderr","time":"2017-12-18T04:35:26.608727757Z"}
{"log":"E 1218 04:35:26.609 UTC THREAD23: service failure: Failure(connection timed out: /100.97.4.75:3000 at remote address: /100.97.4.75:3000. Remote Info: Not Available, flags=0x09) with RemoteInfo -\u003e Upstream Address: Not Available, Upstream id: Not Available, Downstream Address: /100.97.4.75:3000, Downstream label: #/io.l5d.k8s/integrations-staging/http/inventory, Trace Id: 7f31992890b7e429.7f31992890b7e429\u003c:7f31992890b7e429\n","stream":"stderr","time":"2017-12-18T04:35:26.610478624Z"}
```
This is what we've seen before. They deployed this service and `100.97.4.75` is the old endpoint. It continued to try and talk to that endpoint for the next 3 hours, throwing a bunch of ConnectionFailedException errors.

<img width="602" alt="screenshot" src="https://user-images.githubusercontent.com/408/34123769-888f8532-e3e5-11e7-868a-a49782f9409a.png">

It finally recovered on its own, with no work from any person. I found this in the logs for the time when the connection errors stopped (the Grafana graphs are PST and the logs are UTC):
```
{"log":"I 1218 06:48:41.142 UTC THREAD10: Reaping #/io.l5d.k8s/integrations-staging/http/inventory\n","stream":"stderr","time":"2017-12-18T06:48:41.147723284Z"}
```
After this, it picked up the new endpoints and has been working happily.

Hopefully, that helps. I will try running the new version, and also will pay attention to issue #1755 as well.

**Comment by hawkw** [comment]
Hi again @dpetersen, thanks for all the additional information! Hopefully the fixes in 1.3.4 will help, but if not, we can also cut you a snapshot release with the changes in #1755. And if _that_ doesn't work, the information you've provided will help as we continue debugging this issue.

**Comment by dpetersen** [comment]
Bad news, the issue occurred again tonight and we're running v1.3.4. The metrics and logs for this event are identical to what I've already posted, so I won't post them again.

**Comment by hawkw** [comment]
Okay, thanks for keeping us informed. Would you be interested in trying a snapshot release while we continue investigating this issue?

**Comment by dpetersen** [comment]
Sure, we're willing to try a snapshot if it has potential fixes for this issue.

**Comment by hawkw** [comment]
Alright, should I publish a snapshot Docker image or send you a binary? We don't have a working reproduction of the issue you're seeing yet, so we can't yet verify that #1755 definitely fixes it, but the symptoms seem similar.

**Comment by dpetersen** [comment]
Docker image would be the easiest, thanks!

**Comment by hawkw** [comment]
Okay, I've pushed a new Linkerd image tagged as `buoyantio/linkerd:1.3.4-SNAPSHOT-k8s-json-fix`. Please let us know if you continue seeing issues with this snapshot!

**Comment by dpetersen** [comment]
Unfortunately, it didn't take long to get a failure. We've had two failed deployments with the usual symptoms in the last hour since I've deployed. I would say that the move from 1.3.3 to 1.3.4 has exacerbated the problem. We went weeks without any issues after the initial failure, and in the last 24 hours (since first upgrading to 1.3.4 and then to this branch) we've had 6.

All that said, whatever triggers this behavior must be something special about one of our services. I think we've had 8 total failures and 7 of them have been with the same app. Off the top of my head, I know this app is unreliable. It regularly gets inundated with requests, even on staging, so multiple times a day it will be briefly removed from linkerd's internal load balancer due to failure accrual. That's not desirable, but in that case, linkerd behaves as I expect. Our other services are usually less flakey.

I will take a look at recent development on that service to make sure that our increased failure rate isn't something new in the service instead of our 1.3.4 update. The deployment issues still seem like misbehavior from linkerd, but it must be something special we're doing here that triggers it. There were a dozen successful deploys today for other services.

**Comment by hawkw** [comment]
> Unfortunately, it didn't take long to get a failure.

Just to confirm, this is with the snapshot?

**Comment by dpetersen** [comment]
Yes, the two most recent failures are running the snapshot.

**Comment by hawkw** [comment]
> Yes, the two most recent failures are running the snapshot.

Okay, just making sure. I guess that at least we now know the issue you're seeing is different from the one @obeattie describes in #1755. 

I'm sorry you're continuing to see these failures! While it definitely seems like a Linkerd bug is involved, your theory that something about your service is interacting with Linkerd in some way that triggers it is definitely likely from my perspective, especially given that I still haven't been able to reproduce these issues on my end.

**Comment by obeattie** [comment]
Just one more thing to rule out: you definitely have [probation](https://linkerd.io/config/1.3.4/linkerd/index.html#load-balancer) disabled, right?

I don't think you are configuring it explicitly from what you posted above, meaning it should take the default (`false`), but I just wanted to be sure.

**Comment by dpetersen** [comment]
@obeattie I just double-checked and we're not overriding any of the defaults for `enableProbation`, so it should be disabled. Thanks for looking!

**Comment by hollinwilkins** [comment]
Hey,

I am seeing an issue similar to this, or perhaps it is different. For some reason, it is unclear why, certain linkerd instances will not see any endpoints are available for a service.

Here is the error I get:

```
com.twitter.finagle.NoBrokersAvailableException: No hosts are available for /svc/aai-account.default, Dtab.base=[/ph=>/$/io.buoyant.rinet;/svc=>/ph/80;/svc=>/$/io.buoyant.porthostPfx/ph;/k8s=>/#/io.l5d.k8s.h2;/portNsSvc=>/#/portNsSvcToK8s;/host=>/portNsSvc/h2/default;/host=>/portNsSvc/h2;/svc=>/$/io.buoyant.http.domainToPathPfx/host], Dtab.local=[]. Remote Info: Not Available
```

I only get this error on a single instance of linkerd from the daemonset, all other instances are able to communicate with `aai-account.default`.

The `aai-account` definitely has endpoints:

```
kubectl get endpoints -l app=account
NAME          ENDPOINTS          AGE
aai-account   10.52.5.217:5560   9m
```

servicemesh.yaml comes straight from the tutorial, except I enable zipkin and upgrade image to 1.3.4

```
################################################################################
# Linkerd Service Mesh
#
# This is a basic Kubernetes config file to deploy a service mesh of Linkerd
# instances onto your Kubernetes cluster that is capable of handling HTTP,
# HTTP/2 and gRPC calls with some reasonable defaults.
#
# To configure your applications to use Linkerd for HTTP traffic you can set the
# `http_proxy` environment variable to `$(NODE_NAME):4140` where `NODE_NAME` is
# the name of node on which the application instance is running.  The
# `NODE_NAME` environment variable can be set with the downward API.
#
# If your application does not support the `http_proxy` environment variable or
# if you want to configure your application to use Linkerd for HTTP/2 or gRPC
# traffic, you must configure your application to send traffic directly to
# Linkerd:
#
# * $(NODE_NAME):4140 for HTTP
# * $(NODE_NAME):4240 for HTTP/2
# * $(NODE_NAME):4340 for gRPC
#
# If you are sending HTTP or HTTP/2 traffic directly to Linkerd, you must set
# the Host/Authority header to `<service>` or `<service>.<namespace>` where
# `<service>` and `<namespace>` are the names of the service and namespace
# that you want to proxy to.  If unspecified, `<namespace>` defaults to
# `default`.
#
# If your application receives HTTP, HTTP/2, and/or gRPC traffic it must have a
# Kubernetes Service object with ports named `http`, `h2`, and/or `grpc`
# respectively.
#
# You can deploy this to your Kubernetes cluster by running:
#   kubectl create ns linkerd
#   kubectl apply -n linkerd -f servicemesh.yml
#
# There are sections of this config that can be uncommented to enable:
# * CNI compatibility
# * Automatic retries
# * Zipkin tracing
################################################################################
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: l5d-config
  namespace: linkerd
data:
  config.yaml: |-
    admin:
      ip: 0.0.0.0
      port: 9990

    # Namers provide Linkerd with service discovery information.  To use a
    # namer, you reference it in the dtab by its prefix.  We define 4 namers:
    # * /io.l5d.k8s gets the address of the target app
    # * /io.l5d.k8s.http gets the address of the http-incoming Linkerd router on the target app's node
    # * /io.l5d.k8s.h2 gets the address of the h2-incoming Linkerd router on the target app's node
    # * /io.l5d.k8s.grpc gets the address of the grpc-incoming Linkerd router on the target app's node
    namers:
    - kind: io.l5d.k8s
    - kind: io.l5d.k8s
      prefix: /io.l5d.k8s.http
      transformers:
        # The daemonset transformer replaces the address of the target app with
        # the address of the http-incoming router of the Linkerd daemonset pod
        # on the target app's node.
      - kind: io.l5d.k8s.daemonset
        namespace: linkerd
        port: http-incoming
        service: l5d
        # hostNetwork: true # Uncomment if using host networking (eg for CNI)
    - kind: io.l5d.k8s
      prefix: /io.l5d.k8s.h2
      transformers:
        # The daemonset transformer replaces the address of the target app with
        # the address of the h2-incoming router of the Linkerd daemonset pod
        # on the target app's node.
      - kind: io.l5d.k8s.daemonset
        namespace: linkerd
        port: h2-incoming
        service: l5d
        # hostNetwork: true # Uncomment if using host networking (eg for CNI)
    - kind: io.l5d.k8s
      prefix: /io.l5d.k8s.grpc
      transformers:
        # The daemonset transformer replaces the address of the target app with
        # the address of the grpc-incoming router of the Linkerd daemonset pod
        # on the target app's node.
      - kind: io.l5d.k8s.daemonset
        namespace: linkerd
        port: grpc-incoming
        service: l5d
        # hostNetwork: true # Uncomment if using host networking (eg for CNI)
    - kind: io.l5d.rewrite
      prefix: /portNsSvcToK8s
      pattern: "/{port}/{ns}/{svc}"
      name: "/k8s/{ns}/{port}/{svc}"

    # Telemeters export metrics and tracing data about Linkerd, the services it
    # connects to, and the requests it processes.
    telemetry:
    - kind: io.l5d.prometheus # Expose Prometheus style metrics on :9990/admin/metrics/prometheus
    - kind: io.l5d.recentRequests
      sampleRate: 0.25 # Tune this sample rate before going to production
    - kind: io.l5d.zipkin # Uncomment to enable exporting of zipkin traces
      host: zipkin # Zipkin collector address
      port: 9410
      sampleRate: 1.0 # Set to a lower sample rate depending on your traffic volume

    # Usage is used for anonymized usage reporting.  You can set the orgId to
    # identify your organization or set `enabled: false` to disable entirely.
    usage:
      orgId: linkerd-examples-servicemesh

    # Routers define how Linkerd actually handles traffic.  Each router listens
    # for requests, applies routing rules to those requests, and proxies them
    # to the appropriate destinations.  Each router is protocol specific.
    # For each protocol (HTTP, HTTP/2, gRPC) we define an outgoing router and
    # an incoming router.  The application is expected to send traffic to the
    # outgoing router which proxies it to the incoming router of the Linkerd
    # running on the target service's node.  The incoming router then proxies
    # the request to the target application itself.  We also define HTTP and
    # HTTP/2 ingress routers which act as Ingress Controllers and route based
    # on the Ingress resource.
    routers:
    - label: http-outgoing
      protocol: http
      servers:
      - port: 4140
        ip: 0.0.0.0
      # This dtab looks up service names in k8s and falls back to DNS if they're
      # not found (e.g. for external services). It accepts names of the form
      # "service" and "service.namespace", defaulting the namespace to
      # "default". For DNS lookups, it uses port 80 if unspecified. Note that
      # dtab rules are read bottom to top. To see this in action, on the Linkerd
      # administrative dashboard, click on the "dtab" tab, select "http-outgoing"
      # from the dropdown, and enter a service name like "a.b". (Or click on the
      # "requests" tab to see recent traffic through the system and how it was
      # resolved.)
      dtab: |
        /ph  => /$/io.buoyant.rinet ;                     # /ph/80/google.com -> /$/io.buoyant.rinet/80/google.com
        /svc => /ph/80 ;                                  # /svc/google.com -> /ph/80/google.com
        /svc => /$/io.buoyant.porthostPfx/ph ;            # /svc/google.com:80 -> /ph/80/google.com
        /k8s => /#/io.l5d.k8s.http ;                      # /k8s/default/http/foo -> /#/io.l5d.k8s.http/default/http/foo
        /portNsSvc => /#/portNsSvcToK8s ;                 # /portNsSvc/http/default/foo -> /k8s/default/http/foo
        /host => /portNsSvc/http/default ;                # /host/foo -> /portNsSvc/http/default/foo
        /host => /portNsSvc/http ;                        # /host/default/foo -> /portNsSvc/http/default/foo
        /svc => /$/io.buoyant.http.domainToPathPfx/host ; # /svc/foo.default -> /host/default/foo
      client:
        kind: io.l5d.static
        configs:
        # Use HTTPS if sending to port 443
        - prefix: "/$/io.buoyant.rinet/443/{service}"
          tls:
            commonName: "{service}"

    - label: http-incoming
      protocol: http
      servers:
      - port: 4141
        ip: 0.0.0.0
      interpreter:
        kind: default
        transformers:
        - kind: io.l5d.k8s.localnode
          # hostNetwork: true # Uncomment if using host networking (eg for CNI)
      dtab: |
        /k8s => /#/io.l5d.k8s ;                           # /k8s/default/http/foo -> /#/io.l5d.k8s/default/http/foo
        /portNsSvc => /#/portNsSvcToK8s ;                 # /portNsSvc/http/default/foo -> /k8s/default/http/foo
        /host => /portNsSvc/http/default ;                # /host/foo -> /portNsSvc/http/default/foo
        /host => /portNsSvc/http ;                        # /host/default/foo -> /portNsSvc/http/default/foo
        /svc => /$/io.buoyant.http.domainToPathPfx/host ; # /svc/foo.default -> /host/default/foo

    - label: h2-outgoing
      protocol: h2
      experimental: true
      servers:
      - port: 4240
        ip: 0.0.0.0
      dtab: |
        /ph  => /$/io.buoyant.rinet ;                       # /ph/80/google.com -> /$/io.buoyant.rinet/80/google.com
        /svc => /ph/80 ;                                    # /svc/google.com -> /ph/80/google.com
        /svc => /$/io.buoyant.porthostPfx/ph ;              # /svc/google.com:80 -> /ph/80/google.com
        /k8s => /#/io.l5d.k8s.h2 ;                          # /k8s/default/h2/foo -> /#/io.l5d.k8s.h2/default/h2/foo
        /portNsSvc => /#/portNsSvcToK8s ;                   # /portNsSvc/h2/default/foo -> /k8s/default/h2/foo
        /host => /portNsSvc/h2/default ;                    # /host/foo -> /portNsSvc/h2/default/foo
        /host => /portNsSvc/h2 ;                            # /host/default/foo -> /portNsSvc/h2/default/foo
        /svc => /$/io.buoyant.http.domainToPathPfx/host ;   # /svc/foo.default -> /host/default/foo
      client:
        kind: io.l5d.static
        configs:
        # Use HTTPS if sending to port 443
        - prefix: "/$/io.buoyant.rinet/443/{service}"
          tls:
            commonName: "{service}"

    - label: h2-incoming
      protocol: h2
      experimental: true
      servers:
      - port: 4241
        ip: 0.0.0.0
      interpreter:
        kind: default
        transformers:
        - kind: io.l5d.k8s.localnode
          # hostNetwork: true # Uncomment if using host networking (eg for CNI)
      dtab: |
        /k8s => /#/io.l5d.k8s ;                             # /k8s/default/h2/foo -> /#/io.l5d.k8s/default/h2/foo
        /portNsSvc => /#/portNsSvcToK8s ;                   # /portNsSvc/h2/default/foo -> /k8s/default/h2/foo
        /host => /portNsSvc/h2/default ;                    # /host/foo -> /portNsSvc/h2/default/foo
        /host => /portNsSvc/h2 ;                            # /host/default/foo -> /portNsSvc/h2/default/foo
        /svc => /$/io.buoyant.http.domainToPathPfx/host ;   # /svc/foo.default -> /host/default/foo

    - label: grpc-outgoing
      protocol: h2
      experimental: true
      servers:
      - port: 4340
        ip: 0.0.0.0
      identifier:
        kind: io.l5d.header.path
        segments: 1
      dtab: |
        /hp  => /$/inet ;                                # /hp/linkerd.io/8888 -> /$/inet/linkerd.io/8888
        /svc => /$/io.buoyant.hostportPfx/hp ;           # /svc/linkerd.io:8888 -> /hp/linkerd.io/8888
        /srv => /#/io.l5d.k8s.grpc/default/grpc;         # /srv/service/package -> /#/io.l5d.k8s.grpc/default/grpc/service/package
        /svc => /$/io.buoyant.http.domainToPathPfx/srv ; # /svc/package.service -> /srv/service/package
      client:
        kind: io.l5d.static
        configs:
        # Always use TLS when sending to external grpc servers
        - prefix: "/$/inet/{service}"
          tls:
            commonName: "{service}"

    - label: gprc-incoming
      protocol: h2
      experimental: true
      servers:
      - port: 4341
        ip: 0.0.0.0
      identifier:
        kind: io.l5d.header.path
        segments: 1
      interpreter:
        kind: default
        transformers:
        - kind: io.l5d.k8s.localnode
          # hostNetwork: true # Uncomment if using host networking (eg for CNI)
      dtab: |
        /srv => /#/io.l5d.k8s/default/grpc ;             # /srv/service/package -> /#/io.l5d.k8s/default/grpc/service/package
        /svc => /$/io.buoyant.http.domainToPathPfx/srv ; # /svc/package.service -> /srv/service/package

    # HTTP Ingress Controller listening on port 80
    - protocol: http
      label: http-ingress
      servers:
        - port: 80
          ip: 0.0.0.0
          clearContext: true
      identifier:
        kind: io.l5d.ingress
      dtab: /svc => /#/io.l5d.k8s

    # HTTP/2 Ingress Controller listening on port 8080
    - protocol: h2
      experimental: true
      label: h2-ingress
      servers:
        - port: 8080
          ip: 0.0.0.0
          clearContext: true
      identifier:
        kind: io.l5d.ingress
      dtab: /svc => /#/io.l5d.k8s

---
apiVersion: extensions/v1beta1
kind: DaemonSet
metadata:
  labels:
    app: l5d
  name: l5d
  namespace: linkerd
spec:
  template:
    metadata:
      labels:
        app: l5d
    spec:
      # hostNetwork: true # Uncomment to use host networking (eg for CNI)
      volumes:
      - name: l5d-config
        configMap:
          name: "l5d-config"
      containers:
      - name: l5d
        image: buoyantio/linkerd:1.3.4
        env:
        - name: POD_IP
          valueFrom:
            fieldRef:
              fieldPath: status.podIP
        - name: NODE_NAME
          valueFrom:
            fieldRef:
              fieldPath: spec.nodeName
        args:
        - /io.buoyant/linkerd/config/config.yaml
        ports:
        - name: http-outgoing
          containerPort: 4140
          hostPort: 4140
        - name: http-incoming
          containerPort: 4141
        - name: h2-outgoing
          containerPort: 4240
          hostPort: 4240
        - name: h2-incoming
          containerPort: 4241
        - name: grpc-outgoing
          containerPort: 4340
          hostPort: 4340
        - name: grpc-incoming
          containerPort: 4341
        - name: http-ingress
          containerPort: 80
        - name: h2-ingress
          containerPort: 8080
        volumeMounts:
        - name: "l5d-config"
          mountPath: "/io.buoyant/linkerd/config"
          readOnly: true

      # Run `kubectl proxy` as a sidecar to give us authenticated access to the
      # Kubernetes API.
      - name: kubectl
        image: buoyantio/kubectl:v1.8.5
        args:
        - "proxy"
        - "-p"
        - "8001"
---
apiVersion: v1
kind: Service
metadata:
  name: l5d
  namespace: linkerd
spec:
  selector:
    app: l5d
  type: LoadBalancer
  ports:
  - name: http-ingress
    port: 80
  - name: h2-ingress
    port: 8080
```

Kubernetes version: 1.8.4-gke.0

**Comment by bourquep** [comment]
For the record, I am experiencing a similar issue, described in [this thread on Discourse](https://discourse.linkerd.io/t/linkerd-stops-sending-traffic-to-grpc-kubernetes-pods/460/9)

**Comment by pcalcado** [comment]
From @bourquep at the Discourse thread:
> I believe they are the same issue, or at least have the same root cause. What they have in common:
> 
> * The io.l5d.k8s namer stops being able to resolve a k8s service/endpoint.
> * The problem is intermittent, with no obvious trigger (although the OP of the #1730 issue mentions that deploying a new version of a service sometimes triggers the bug, on my end it starts occurring out of the blue - nothing was updated on my test cluster during the Holidays and the problem manifested itself).
> * We both see k8s resource too old in the linkerd logs, accompanied by k8s restarting watch messages.
> * We both see context canceled errors in the kubectl container logs from the l5d pods
> * Restarting l5d (killing the pods and letting the DaemonSet recreate them) fixes the issue.

**Comment by bourquep** [comment]
Some observations:

* I see quite a few `k8s restarting watch on /api/v1/watch/namespaces/studyo/endpoints/generator, resource version None was too old` in the logs, but it seems like these show up even during periods where my service is accessible, so it may not be related to the issue. However,  `resource version None` seems suspicious...

* When the issue happens, I see a ton of these in the logs:
```
retrying k8s request to /api/v1/namespaces/studyo/endpoints/generator on error Failure(Connection refused: localhost/127.0.0.1:8001 at remote address: localhost/127.0.0.1:8001. Remote Info: Not Available, flags=0x08) with RemoteInfo -> Upstream Address: Not Available, Upstream id: Not Available, Downstream Address: localhost/127.0.0.1:8001, Downstream label: client
```
And some (not as many) of these:
```
Exception propagated to the default monitor (upstream address: /10.8.0.168:42958, downstream address: localhost/127.0.0.1:8001, label: client).
```
If I am not mistaken, `127.0.0.1:8001` is the address of the `kubectl proxy`, so it looks like linkerd is unable to communicate with `kubectl`, therefore unable to resolve services/endpoints

* I also see a few of these in the `kubectl` container logs: `httputil: ReverseProxy read error during body copy: context canceled`. But the timestamps don't correlate to the aforementioned connection errors to localhost:8001. 

According to https://github.com/golang/go/issues/20071, this _context canceled_ error happens when a client prematurely abandon a proxied HTTP request. Not sure what to make of this with regards to how linkerd interacts with the k8s proxy, but thought it might be worth posting here for your consideration.

**Comment by siggy** [comment]
Quick update: we are continuing to investigate this issue. Our current test setup includes:
- kubernetes 1.8.4 on GKE
- linkerd configuration based on https://github.com/linkerd/linkerd-examples/blob/master/k8s-daemonset/k8s/servicemesh.yml
- 1 [strest-grpc](https://github.com/BuoyantIO/strest-grpc/) client
- 3 [strest-grpc](https://github.com/BuoyantIO/strest-grpc/) servers, redeployed every 60 seconds
- prometheus + grafana for monitoring

Thus far we have not observed the issue. If anyone on this thread has a tighter reproduction case, we'd love to use it. Ideally a single kubernetes config file that manifests the issue.

For everyone's reference, here is the complete Kubernetes config file with which we are attempting to reproduce this issue:
https://gist.github.com/siggy/9f283c853c3fea959a0007948d6fdcb4

**Comment by bourquep** [comment]
I've been running Linkerd on my production cluster for the past 2 weeks, and so far I had to restart the l5d pods only once (the issue described in this thread - l5d not being able to resolve gRPC endpoints - happened only once). The issue is still happening every few hours/days on my test cluster.

Both clusters are running the same k8s version on GKE, 3 nodes, same Linkerd version, same k8s deployment yaml files. The only difference is that most of my deployments on test have a single pod, whereas I have dozens of pods per service on prod.

Could this issue have something to do with l5d on a specific node not being able to resolve a service which doesn't have any pods running on the l5d node?

**Comment by siggy** [comment]
Pinging this ticket to see if anyone has found a tighter reproduction case. In lieu of that, we are planning to implement #1780, which should give everyone far better insight into linkerd and namerd's internal service discovery and namer states.

**Comment by dpetersen** [comment]
@siggy I don't, sorry. Since I last commented on this I haven't had time to revisit it, and we're continuing to run an older version of Linkerd. Definitely willing to experiment again if a new version has a prospective fix or just additional logging.

**Comment by bourquep** [comment]
@siggy Status quo here. Our dev/test cluster exhibits this issue almost daily, while our prod cluster is much more immune to it (See [my last comment](https://github.com/linkerd/linkerd/issues/1730#issuecomment-357831283))

**Comment by wmorgan** [comment]
@adleong @siggy would #1768 be helpful for diagnosing this?

**Comment by siggy** [comment]
@wmorgan it might.

For those following along at home, when you observe linkerd's routing state to be out of sync with Kubernetes, it would be interesting to observe this discrepancy directly. Some example commands to do this:

```bash
# linkerd 
curl [ip-address-to-bad-linkerd]:9990/client_state.json

# kubernetes
kubectl --all-namespaces=true describe svc
```

**Comment by krak3n** [comment]
I don't know if this is related at all but I have been noticing problems using the HTTP API to update the dtab namesapce using namerd.

Occasionally `PUT` requests return the expected `204 No Content` but do not update the dtab namespace (at least visibile in the admin), restarting the pods then reveals changes to namerd.

So I am guessing this is a problem with watching the CRD, here is my stack trace:

```
W 0129 14:18:51.382 UTC THREAD26 TraceId:dee913496b808c02: Exception propagated to the default monitor (upstream address: /10.32.0.1:46404, downstream address: localhost/127.0.0.1:8001, label: /$/inet/localhost/8001).
io.buoyant.k8s.Watchable$Closed$
        at io.buoyant.k8s.Watchable$Closed$.<clinit>(Watchable.scala)
        at io.buoyant.k8s.Watchable.$anonfun$activity$10(Watchable.scala:284)
        at com.twitter.util.Closable$$anon$5.close(Closable.scala:144)
        at com.twitter.util.Var$$anon$4.$anonfun$closable$1(Var.scala:399)
        at com.twitter.util.Closable$$anon$5.close(Closable.scala:144)
        at com.twitter.util.Closable$.com$twitter$util$Closable$$safeClose(Closable.scala:67)
        at com.twitter.util.Closable$$anon$3.closeSeq(Closable.scala:125)
        at com.twitter.util.Closable$$anon$3.onTry$1(Closable.scala:122)
        at com.twitter.util.Closable$$anon$3.closeSeq(Closable.scala:128)
        at com.twitter.util.Closable$$anon$3.close(Closable.scala:134)
        at com.twitter.util.Closable$.com$twitter$util$Closable$$safeClose(Closable.scala:67)
        at com.twitter.util.Closable$$anon$3.closeSeq(Closable.scala:125)
        at com.twitter.util.Closable$$anon$3.close(Closable.scala:134)
        at com.twitter.util.Closable.close(Closable.scala:20)
        at com.twitter.util.Closable.close$(Closable.scala:20)
        at com.twitter.util.Closable$$anon$3.close(Closable.scala:101)
        at com.twitter.util.Event.$anonfun$toFuture$1(Event.scala:286)
        at com.twitter.util.Future.$anonfun$ensure$1(Future.scala:1549)
        at com.twitter.util.Future.$anonfun$ensure$1$adapted(Future.scala:1548)
        at com.twitter.util.Promise$Monitored.apply(Promise.scala:205)
        at com.twitter.util.Promise$Monitored.apply(Promise.scala:200)
        at com.twitter.util.Promise$$anon$7.run(Promise.scala:532)
        at com.twitter.concurrent.LocalScheduler$Activation.run(Scheduler.scala:198)
        at com.twitter.concurrent.LocalScheduler$Activation.submit(Scheduler.scala:157)
        at com.twitter.concurrent.LocalScheduler.submit(Scheduler.scala:274)
        at com.twitter.concurrent.Scheduler$.submit(Scheduler.scala:109)
        at com.twitter.util.Promise.runq(Promise.scala:522)
        at com.twitter.util.Promise.updateIfEmpty(Promise.scala:880)
        at com.twitter.finagle.netty4.transport.ChannelTransport.com$twitter$finagle$netty4$transport$ChannelTransport$$fail(ChannelTransport.scala:95)
        at com.twitter.finagle.netty4.transport.ChannelTransport$$anon$1.channelInactive(ChannelTransport.scala:188)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:224)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:75)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:224)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:75)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:224)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:75)
        at com.twitter.finagle.netty4.channel.ChannelRequestStatsHandler.channelInactive(ChannelRequestStatsHandler.scala:35)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:224)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:75)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:224)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:75)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:224)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:75)
        at io.netty.handler.codec.MessageAggregator.channelInactive(MessageAggregator.java:417)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:224)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:75)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:224)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:75)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:224)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:75)
        at io.netty.handler.codec.http.HttpContentDecoder.channelInactive(HttpContentDecoder.java:205)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:224)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:75)
        at io.netty.handler.codec.http.HttpContentEncoder.channelInactive(HttpContentEncoder.java:299)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:224)
        at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelInactive(CombinedChannelDuplexHandler.java:420)
        at io.netty.handler.codec.ByteToMessageDecoder.channelInputClosed(ByteToMessageDecoder.java:377)
        at io.netty.handler.codec.ByteToMessageDecoder.channelInactive(ByteToMessageDecoder.java:342)
        at io.netty.channel.CombinedChannelDuplexHandler.channelInactive(CombinedChannelDuplexHandler.java:223)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:224)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:75)
        at com.twitter.finagle.netty4.channel.ChannelStatsHandler.channelInactive(ChannelStatsHandler.scala:131)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:224)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelInactive(DefaultChannelPipeline.java:1337)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231)
        at io.netty.channel.DefaultChannelPipeline.fireChannelInactive(DefaultChannelPipeline.java:916)
        at io.netty.channel.AbstractChannel$AbstractUnsafe$7.run(AbstractChannel.java:744)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:463)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at com.twitter.finagle.util.BlockingTimeTrackingThreadFactory$$anon$1.run(BlockingTimeTrackingThreadFactory.scala:23)
        at java.lang.Thread.run(Thread.java:748)

E 0129 14:18:51.427 UTC THREAD26: retrying k8s request to /apis/l5d.io/v1alpha1/namespaces/my-namespace/dtabs on error io.buoyant.k8s.Watchable$Closed$
W 130 14:18:51.516 UTC THREAD14 TraceId:44f1d79477bfaaa5: k8s failed to get resource at /apis/l5d.io/v1alpha1/namespaces/my-namespace/dtabs: Failure(io.buoyant.k8s.Watchable$Closed$, flags=0x02) with RemoteInfo -> Upstream Address: Not Available, Upstream id: Not Available, Downstream Address: localhost/127.0.0.1:8001, Downstream label: /$/inet/localhost/8001, Trace Id: 7f9a5b33c94d0c63.7f9a5b33c94d0c63<:7f9a5b33c94d0c63
```

**Comment by deebo91** [comment]
@krak3n your issue may be related but I am not sure. Could you give us more information about the environment namerd is running in? I am assuming you are using k8s 1.8 since you referred to CRDs. What version of Namerd are you using?

**Comment by krak3n** [comment]
@deebo91 yup I am running in Kubernetets `1.8.5` in Google Kubernetes Engine and running namerd `1.3.5` with a kubectl sidecar at version `1.8.5`.

Here is the configuration for `namerd` I am using:

``` yaml
    # Admin UI
    admin:
      ip: 0.0.0.0
      port: 9991
    # DTab Storage location (k8s custom resource definition)
    storage:
      kind: io.l5d.k8s
      namespace: my-namespace
      experimental: true
    # Namers
    namers:
      # Kubernetes Namer
      - kind: io.l5d.k8s
        host: 127.0.0.1
        port: 8001
    # External Interfaces to namerd
    interfaces:
      - kind: io.l5d.thriftNameInterpreter
        ip: 0.0.0.0
        port: 4100
      - kind: io.l5d.httpController
        ip: 0.0.0.0
        port: 4180
      - kind: io.l5d.mesh
        ip: 0.0.0.0
        port: 4321
```

Here is the full Deployment for `namerd` (via `helm`):

``` yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "1"
  creationTimestamp: 2018-01-29T15:41:39Z
  generation: 1
  labels:
    app: namerd-namerd
    chart: namerd-2018.1.29
    heritage: Tiller
    release: namerd
  name: namerd-namerd
  namespace: my-namespace
  resourceVersion: "4079447"
  selfLink: /apis/extensions/v1beta1/namespaces/my-namespace/deployments/namerd-namerd
  uid: e5a93b3d-050a-11e8-bddc-42010a9a000c
spec:
  replicas: 3
  selector:
    matchLabels:
      app: namerd-namerd
      release: namerd
  strategy:
    rollingUpdate:
      maxSurge: 1
      maxUnavailable: 1
    type: RollingUpdate
  template:
    metadata:
      annotations:
        checksum/config: d856cc693b32c2fd5cec2f6be7329b817f13f018d16083cd951ebda26778e72c
      creationTimestamp: null
      labels:
        app: namerd-namerd
        release: namerd
    spec:
      containers:
      - args:
        - /io.buoyant/namerd/config/config.yaml
        image: buoyantio/namerd:1.3.5
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /api/1/dtabs
            port: 4180
            scheme: HTTP
          initialDelaySeconds: 60
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        name: namerd-namerd
        ports:
        - containerPort: 4100
          name: sync
          protocol: TCP
        - containerPort: 4180
          name: api
          protocol: TCP
        - containerPort: 4321
          name: mesh
          protocol: TCP
        - containerPort: 9991
          name: admin
          protocol: TCP
        readinessProbe:
          failureThreshold: 3
          httpGet:
            path: /api/1/dtabs
            port: 4180
            scheme: HTTP
          initialDelaySeconds: 60
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        resources:
          limits:
            cpu: 500m
            memory: 512Mi
          requests:
            cpu: "0"
            memory: 512Mi
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /io.buoyant/namerd/config
          name: namerd-namerd-config
          readOnly: true
      - args:
        - proxy
        - -p
        - "8001"
        image: buoyantio/kubectl:v1.8.5
        imagePullPolicy: IfNotPresent
        name: kubectl
        resources:
          requests:
            cpu: "0"
            memory: 32Mi
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: namerd
      serviceAccountName: namerd
      terminationGracePeriodSeconds: 30
      volumes:
      - configMap:
          defaultMode: 420
          name: namerd-namerd-config
        name: namerd-namerd-config
status:
  availableReplicas: 3
  conditions:
  - lastTransitionTime: 2018-01-29T15:42:51Z
    lastUpdateTime: 2018-01-29T15:42:51Z
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  observedGeneration: 1
  readyReplicas: 3
  replicas: 3
  updatedReplicas: 3
```

**Comment by deebo91** [comment]
Thanks @krak3n, everything in your configs look fine to me. Do you think you could create a small docker setup that exhibits the issues you mentioned. It could help us pinpoint were exactly were the error is happening. Also, I think I am going to create a separate issue as I think this may not be related to k8s talking to old endpoints.

**Comment by pavel-mikhalchuk** [comment]
Hey!

Just wanted to share some details on issue I had couple minutes ago. Unfortunately I didn't grab logs :(

After another deployment of one of my apps in some namespace linkerd stopped serving requests. 

I have restarted linkerd and after some time I got this exception:

`E 0207 09:36:27.703 UTC THREAD16: retrying k8s request to /api/v1/namespaces/dev-dev/endpoints/s-portal-m-server on error com.twitter.io.Reader$ReaderDiscarded: This writer's reader has been discarded
W 0207 09:36:27.718 UTC THREAD16 TraceId:adb9457471620bb5: Exception propagated to the default monitor (upstream address: /10.100.20.21:59600, downstream address: localhost/127.0.0.1:7001, label: client).
com.twitter.io.Reader$ReaderDiscarded: This writer's reader has been discarded
	at com.twitter.finagle.netty4.http.StreamTransports$$anon$1.discard(StreamTransports.scala:71)
	at com.twitter.finagle.http.DelayedReleaseService$$anon$2$$anon$3.discard(DelayedReleaseService.scala:58)
	at com.twitter.finagle.http.DelayedReleaseService$$anon$2$$anon$3.discard(DelayedReleaseService.scala:58)
	at io.buoyant.k8s.Watchable.$anonfun$watch$5(Watchable.scala:131)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at com.twitter.util.Try$.apply(Try.scala:15)
	at com.twitter.util.Future$.apply(Future.scala:166)
	at io.buoyant.k8s.Watchable.$anonfun$watch$4(Watchable.scala:131)`

But if I **curl localhost:7001/api/v1/namespaces/dev-dev/endpoints/s-portal-m-server** manually on the server where linkerd is running (hostNetwork is set to true) I get:

`{
  "kind": "Endpoints",
  "apiVersion": "v1",
  "metadata": {
    "name": "s-portal-m-server",
    "namespace": "dev-dev",
    "selfLink": "/api/v1/namespaces/dev-dev/endpoints/s-portal-m-server",
    "uid": "e3a0e4dc-c156-11e7-bbdc-005056b554ea",
    "resourceVersion": "17002723",
    "creationTimestamp": "2017-11-04T11:54:18Z",
    "labels": {
      "env": "dev-dev",
      "app": "portal-m"
    }
  },
  "subsets": [
    {
      "addresses": [
        {
          "ip": "10.44.0.18",
          "nodeName": "asvc-tst-pool1",
          "targetRef": {
            "kind": "Pod",
            "namespace": "dev-dev",
            "name": "d-portal-m-server-5b764dfb67-nphbf",
            "uid": "839dcc96-0b56-11e8-bbdc-005056b554ea",
            "resourceVersion": "17002722"
          }
        }
      ],
      "ports": [
        {
          "name": "http",
          "port": 8030,
          "protocol": "TCP"
        }
      ]
    }
  ]
}`

Linkerd **kubectl** sidecar is running on 7001. Could it be **kubectl** **proxy** an issue? I remember port forwarding is not working very stable and I'm getting 404 from time to time. 

Thanks!

**Comment by hawkw** [comment]
Just wanted to give you all a little update: we've finally seen this type of failure in our test cluster, and that's helped to provide a little more insight into where the bug might be. I'm still investigating the issue, but there has been some progress!

**Comment by bourquep** [comment]
@hawkw Oh, that is interesting. We delete a lot of resources on our test cluster, because we have "review" deployments triggered by CI for each branch of each service, and once the branch is merged we delete the relevant k8s resources (service, deployment).

So this might explain why I am experiencing this issue almost daily on my test cluster, but almost never on my prod cluster, where k8s resources are updated but seldom deleted.

**Comment by hawkw** [comment]
@bourquep in that case, it sounds very likely that the issue you're experiencing is in fact #1791. Fortunately, we have a fix for that bug, as well -- if you'd like, I can publish a snapshot image for that issue as well. 

If you do switch to a build with the fix for #1791 and continue to see occasional failures, it's likely that any remaining routing failures are caused by _this_ bug (#1730) --- this may be the case if you are seeing intermittent routing failures on the prod cluster despite never deleting k8s resources. However, I would expect that a build with the fix for incorrect DELETED event handling will lead to a significantly reduced rate of incorrect routing in your test cluster.

**Comment by bourquep** [comment]
@hawkw Unfortunately, I have replaced Linkerd with Conduit on my test cluster, which is where I was experiencing this issue most of the time. If you feel like it would really help if I tested a new Linkerd version, I could probably bring it back into that cluster, but I'm kinda in love with Conduit now and I'd rather stick with it... ;)

**Comment by hawkw** [comment]
@bourquep no, that's quite alright! I'm happy to hear that your Conduit adoption has been so successful! :D

**Comment by adleong** [comment]
@dpetersen are you still seeing this issue with Linkerd 1.3.6

**Comment by dpetersen** [comment]
Honestly, I have been so busy we haven't been able to make another upgrade attempt. We're still running the same version we have been running. Not sure when I'm going to be able to revisit it. I'm fine if you think it's fixed and want to close this issue. I can re-open or open something new if it's not fixed.

**Comment by adleong** [comment]
Thanks, @dpetersen.  Hard to say for sure if this is fixed or not given our inability to reproduce it.  I'm going to close this issue for now but please re-open if you see this again after upgrading.

**Comment by yoitsro** [comment]
I'm still seeing this on 1.3.7. Can anyone recommend a short term fix/workaround in the meantime?

Is it likely this could be fixed in 1.4.0?

**Comment by yoitsro** [comment]
Strangely enough, the disk usage seems to creep up and get to a certain threshold when eventually, it just stops serving all traffic:

![image](https://user-images.githubusercontent.com/3006852/39672895-b1e10804-512a-11e8-88f9-d486b9287e56.png)
![image](https://user-images.githubusercontent.com/3006852/39672904-d0606af4-512a-11e8-8f0a-22f1436e1025.png)
![image](https://user-images.githubusercontent.com/3006852/39672906-d978889c-512a-11e8-8ae2-21b1eaa2e5e6.png)

Are these just logs being stored to disk somewhere, which is what's causing the increased disk usage?

**Comment by yoitsro** [comment]
Anyone? :)

**Comment by ejwood79** [comment]
We actually just stopped using linkerd altogether due to this issue.

**Comment by yoitsro** [comment]
@ejwood79 I'm about to jump ship as well. What are you using now?

**Comment by ejwood79** [comment]
Starting to look at istio.

**Comment by dadjeibaah** [comment]
@yoitsro, it is a bummer that you are running into this. Since this issue is not easy to reproduce, we have added new functionality that inspects the state of the k8s namer. We would love to see how we could use this new feature to debug this issue if that is something you are willing to try. the k8s namer inspector has not been released yet but is available on latest master. I believe there is a nightly linkerd docker image you could use to diagnose the issue. If there is any additional information you may have that might help resolve this, please do let us know!

**Comment by wmorgan** [comment]
@ejwood79 are you running into an increase in disk usage? Or into Linkerd talking to old K8s endpoints? If so, can you clarify which version of Linkerd you are running, and which version of K8s?

**Comment by ejwood79** [comment]
It was the issue with linkerd talking to pods that are no longer scheduled. We've since removed linkerd from our cluster, though. It was on kubernetes 1.7.

**Comment by adleong** [comment]
Hi @yoitsro, very sorry that no one responded to you until today.  I must have missed your comments because they were on a closed issue.  This thread, unfortunately, conflates several reports which may or may not have the same root cause.  

@yoitsro and @ejwood79, if either of you are still interested in getting to the bottom of this, we're more than happy to help you debug.  If you could open new issues with as much detail as possible about your environment (Linkerd version, config, etc) and the symptoms you're seeing, that will help us to dig in and figure out what's going on.

I'm going to re-close this particular issue because the original issue has already been fixed.  But I hope you do file new issues so that we can investigate.  Sorry, again, for the lack of communication and thanks for your understanding.
